File: 2801. Count Stepping Numbers in Range/Solution.py
class Solution:
    def countSteppingNumbers(self, low: str, high: str) -> int:
        self.dp = [[1 if i == 1 else 0 for j in range(10)] for i in range(101)]
        self.mod = 1000000007
        return (self.count(high) - self.count(low) + self.check(low)) % self.mod

    def check(self, x):
        for i in range(1, len(x)):
            if abs(int(x[i]) - int(x[i - 1])) != 1: return 0
        return 1

    def count(self, x):
        res = 0
        for i in range(1, len(x) + 1):
            for d in range(1, 10):
                res += self.dfs(i, d, x, i == len(x))
        return res

    def dfs(self, i, d, x, lim):
        if d > 9 or d < 0: return 0
        if lim and d > int(x[len(x) - i]): return 0
        if i == 1: return self.dp[i][d]
        lim = lim and (d == int(x[len(x) - i]))
        if lim:
            return (self.dfs(i - 1, d - 1, x, True) + self.dfs(i - 1, d + 1, x, True)) % self.mod
        if self.dp[i][d] != 0: return self.dp[i][d] - 1
        self.dp[i][d] = (1 + self.dfs(i - 1, d - 1, x, False) + self.dfs(i - 1, d + 1, x, False)) % self.mod
        return self.dp[i][d] - 1

File: 2801. Count Stepping Numbers in Range/Solution.cpp
class Solution {
public:
    vector<vector<long long>> dp;
    long long mod = 1000000007;
    int countSteppingNumbers(string low, string high) {
        dp.assign(101, vector<long long>(10, 0));
        for (int i = 0; i < 10; i++) dp[1][i] = 1;
        return (int) ((mod + count(high) - count(low) + check(low)) % mod);
    }

    long long check(string x) {
        for (int i = 1; i < x.length(); i++) {
            if (abs(x[i] - x[i - 1]) != 1) return 0;
        }
        return 1;
    }

    long long count(string x) {
        long long res = 0;
        for (int i = 1; i < x.length() + 1; i++) {
            for (int d = 1; d < 10; d++) {
                res = (res + dfs(i, d, x, i == x.length())) % mod;
            }
        }
        return res;
    }

    long long dfs(int i, int d, string x, bool lim) {
        if (d < 0 || d > 9) return 0;
        if (lim && (d > x[x.length() - i] - '0')) return 0;
        if (i == 1) return dp[i][d];
        lim = (lim && (d == x[x.length() - i] - '0'));
        if (lim) return (dfs(i - 1, d - 1, x, true) + dfs(i - 1, d + 1, x, true)) % mod;
        if (dp[i][d] != 0) return dp[i][d] - 1LL;
        dp[i][d] = (1LL + dfs(i - 1, d - 1, x, false) + dfs(i - 1, d + 1, x, false)) % mod;
        return dp[i][d] - 1LL;
    }
};
